File: app/src/main/java/jp/juggler/subwaytooter/view/GravitySnapHelper.java
package jp.juggler.subwaytooter.view;

import android.annotation.SuppressLint;
import android.graphics.PointF;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.LinearSnapHelper;
import android.support.v7.widget.OrientationHelper;
import android.support.v7.widget.RecyclerView;
import android.view.Gravity;
import android.view.View;
import android.view.animation.DecelerateInterpolator;
import android.widget.Scroller;

import jp.juggler.subwaytooter.util.LogCategory;

public class GravitySnapHelper extends LinearSnapHelper {
	
	static final LogCategory log = new LogCategory( "GravitySnapHelper" );
	
	private OrientationHelper verticalHelper;
	private OrientationHelper horizontalHelper;
	private int gravity;
	private boolean isRTL;
	
	private RecyclerView mRecyclerView;
	private Scroller mGravityScroller;
	
	@SuppressLint("RtlHardcoded")
	public GravitySnapHelper( int gravity ){
		this.gravity = gravity;
		if( this.gravity == Gravity.LEFT ){
			this.gravity = Gravity.START;
		}else if( this.gravity == Gravity.RIGHT ){
			this.gravity = Gravity.END;
		}
	}
	
	@Override
	public void attachToRecyclerView( @Nullable RecyclerView recyclerView )
		throws IllegalStateException{
		mRecyclerView = recyclerView;
		if( recyclerView != null ){
			isRTL = ViewCompat.getLayoutDirection( recyclerView ) == ViewCompat.LAYOUT_DIRECTION_RTL;
			
			mGravityScroller = new Scroller(mRecyclerView.getContext(), new DecelerateInterpolator());
		}
		super.attachToRecyclerView( recyclerView );
	}
	
	@Override public int[] calculateDistanceToFinalSnap(
		@NonNull RecyclerView.LayoutManager layoutManager
		, @NonNull View targetView
	){
		int[] out = new int[ 2 ];
		
		if( ! layoutManager.canScrollHorizontally() ){
			out[ 0 ] = 0;
		}else if( gravity == Gravity.START ){
			out[ 0 ] = distanceToStart( targetView, getHorizontalHelper( layoutManager ) );
		}else{
			out[ 0 ] = distanceToEnd( targetView, getHorizontalHelper( layoutManager ) );
		}
		
		if( ! layoutManager.canScrollVertically() ){
			out[ 1 ] = 0;
		}else if( gravity == Gravity.TOP ){
			out[ 1 ] = distanceToStart( targetView, getVerticalHelper( layoutManager ) );
		}else{
			out[ 1 ] = distanceToEnd( targetView, getVerticalHelper( layoutManager ) );
		}
		
		return out;
	}
	
	@Override public View findSnapView( RecyclerView.LayoutManager layoutManager ){
		if( layoutManager instanceof LinearLayoutManager ){
			switch( gravity ){
			
			case Gravity.START:
				return findStartView( layoutManager, getHorizontalHelper( layoutManager ) );
			
			case Gravity.TOP:
				return findStartView( layoutManager, getVerticalHelper( layoutManager ) );
			
			case Gravity.END:
				return findEndView( layoutManager, getHorizontalHelper( layoutManager ) );
			
			case Gravity.BOTTOM:
				return findEndView( layoutManager, getVerticalHelper( layoutManager ) );
			}
		}
		
		return super.findSnapView( layoutManager );
	}
	
	private int distanceToStart( View targetView, OrientationHelper helper ){
		if( isRTL ){
			return helper.getDecoratedEnd( targetView ) - helper.getEndAfterPadding();
		}else{
			return helper.getDecoratedStart( targetView ) - helper.getStartAfterPadding();
		}
	}
	
	private int distanceToEnd( View targetView, OrientationHelper helper ){
		if( isRTL ){
			return helper.getDecoratedStart( targetView ) - helper.getStartAfterPadding();
		}else{
			return helper.getDecoratedEnd( targetView ) - helper.getEndAfterPadding();
		}
	}
	
	private View findStartView(
		RecyclerView.LayoutManager layoutManager
		, OrientationHelper helper
	){
		
		if( layoutManager instanceof LinearLayoutManager ){
			LinearLayoutManager llm = (LinearLayoutManager) layoutManager;
			
			int firstChild = llm.findFirstVisibleItemPosition();
			
			if( firstChild == RecyclerView.NO_POSITION ){
				return null;
			}
			
			View child = layoutManager.findViewByPosition( firstChild );
			
			if( helper.getDecoratedEnd( child ) >= helper.getDecoratedMeasurement( child ) / 2
				&& helper.getDecoratedEnd( child ) > 0
				){
				return child;
			}else if( llm.findLastCompletelyVisibleItemPosition() == layoutManager.getItemCount() - 1 ){
				return null;
			}else{
				return layoutManager.findViewByPosition( firstChild + 1 );
			}
		}
		
		return super.findSnapView( layoutManager );
	}
	
	private View findEndView(
		RecyclerView.LayoutManager layoutManager
		, OrientationHelper helper
	){
		
		if( layoutManager instanceof LinearLayoutManager ){
			LinearLayoutManager llm = (LinearLayoutManager) layoutManager;
			
			int lastChild = llm.findLastVisibleItemPosition();
			
			if( lastChild == RecyclerView.NO_POSITION ){
				return null;
			}
			
			View child = llm.findViewByPosition( lastChild );
			
			if( helper.getDecoratedStart( child ) + helper.getDecoratedMeasurement( child ) / 2
				<= helper.getTotalSpace() ){
				return child;
			}else if( llm.findFirstCompletelyVisibleItemPosition() == 0 ){
				return null;
			}else{
				return llm.findViewByPosition( lastChild - 1 );
			}
		}
		
		return super.findSnapView( layoutManager );
	}
	
	private OrientationHelper getVerticalHelper( RecyclerView.LayoutManager layoutManager ){
		if( verticalHelper == null ){
			verticalHelper = OrientationHelper.createVerticalHelper( layoutManager );
		}
		return verticalHelper;
	}
	
	private OrientationHelper getHorizontalHelper( RecyclerView.LayoutManager layoutManager ){
		if( horizontalHelper == null ){
			horizontalHelper = OrientationHelper.createHorizontalHelper( layoutManager );
		}
		return horizontalHelper;
	}
	
	int column_w;
	
	public void setColumnWidth( int column_w ){
		this.column_w = column_w;
	}

	@Override public int findTargetSnapPosition(
		RecyclerView.LayoutManager layoutManager
		, int velocityX
		, int velocityY
	) {
		
		int targetPos = super.findTargetSnapPosition( layoutManager,velocityX,velocityY );
		if( targetPos != RecyclerView.NO_POSITION ){
			final View currentView = findSnapView(layoutManager);
			if( currentView != null){
				final int currentPosition = layoutManager.getPosition( currentView );
				
				int clip = 1;
				
				if( targetPos - currentPosition > clip ){
					targetPos = currentPosition + clip;
				}else if( targetPos - currentPosition < - clip ){
					targetPos = currentPosition - clip;
				}
			}
			
		}
		return targetPos;
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/view/GravitySnapHelper.kt
package jp.juggler.subwaytooter.view

import android.annotation.SuppressLint
import android.view.Gravity
import android.view.View
import android.view.animation.DecelerateInterpolator
import android.widget.Scroller
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView
import jp.juggler.util.log.LogCategory

class GravitySnapHelper @SuppressLint("RtlHardcoded")
constructor(gravity: Int) : androidx.recyclerview.widget.LinearSnapHelper() {

    companion object {

        internal val log = LogCategory("GravitySnapHelper")
    }

    private var verticalHelper: OrientationHelper? = null
    private var horizontalHelper: OrientationHelper? = null
    private var gravity: Int = 0
    private var isRTL: Boolean = false

    private var mRecyclerView: RecyclerView? = null
    private var mGravityScroller: Scroller? = null

    init {
        this.gravity = gravity
        if (this.gravity == Gravity.LEFT) {
            this.gravity = Gravity.START
        } else if (this.gravity == Gravity.RIGHT) {
            this.gravity = Gravity.END
        }
    }

    @Throws(IllegalStateException::class)
    override fun attachToRecyclerView(recyclerView: RecyclerView?) {
        mRecyclerView = recyclerView
        if (recyclerView != null) {
            isRTL = recyclerView.getLayoutDirection() == View.LAYOUT_DIRECTION_RTL

            mGravityScroller = Scroller(recyclerView.context, DecelerateInterpolator())
        }
        super.attachToRecyclerView(recyclerView)
    }

    override fun calculateDistanceToFinalSnap(
        layoutManager: RecyclerView.LayoutManager,
        targetView: View,
    ): IntArray {
        val out = IntArray(2)

        if (!layoutManager.canScrollHorizontally()) {
            out[0] = 0
        } else if (gravity == Gravity.START) {
            out[0] = distanceToStart(targetView, getHorizontalHelper(layoutManager))
        } else {
            out[0] = distanceToEnd(targetView, getHorizontalHelper(layoutManager))
        }

        if (!layoutManager.canScrollVertically()) {
            out[1] = 0
        } else if (gravity == Gravity.TOP) {
            out[1] = distanceToStart(targetView, getVerticalHelper(layoutManager))
        } else {
            out[1] = distanceToEnd(targetView, getVerticalHelper(layoutManager))
        }

        return out
    }

    override fun findSnapView(layoutManager: RecyclerView.LayoutManager): View? {
        if (layoutManager is LinearLayoutManager) {
            when (gravity) {

                Gravity.START -> return findStartView(
                    layoutManager,
                    getHorizontalHelper(layoutManager)
                )

                Gravity.TOP -> return findStartView(layoutManager, getVerticalHelper(layoutManager))

                Gravity.END -> return findEndView(layoutManager, getHorizontalHelper(layoutManager))

                Gravity.BOTTOM -> return findEndView(
                    layoutManager,
                    getVerticalHelper(layoutManager)
                )
            }
        }

        return super.findSnapView(layoutManager)
    }

    private fun distanceToStart(targetView: View, helper: OrientationHelper): Int {
        return if (isRTL) {
            helper.getDecoratedEnd(targetView) - helper.endAfterPadding
        } else {
            helper.getDecoratedStart(targetView) - helper.startAfterPadding
        }
    }

    private fun distanceToEnd(targetView: View, helper: OrientationHelper): Int {
        return if (isRTL) {
            helper.getDecoratedStart(targetView) - helper.startAfterPadding
        } else {
            helper.getDecoratedEnd(targetView) - helper.endAfterPadding
        }
    }

    private fun findStartView(
        layoutManager: RecyclerView.LayoutManager,
        helper: OrientationHelper,
    ): View? {

        if (layoutManager is LinearLayoutManager) {

            val firstChild = layoutManager.findFirstVisibleItemPosition()

            if (firstChild == RecyclerView.NO_POSITION) {
                return null
            }

            val child = layoutManager.findViewByPosition(firstChild)

            @Suppress("CascadeIf")
            return if (helper.getDecoratedEnd(child) >= helper.getDecoratedMeasurement(child) / 2 && helper.getDecoratedEnd(
                    child
                ) > 0
            ) {
                child
            } else if (layoutManager.findLastCompletelyVisibleItemPosition() == layoutManager.getItemCount() - 1) {
                null
            } else {
                layoutManager.findViewByPosition(firstChild + 1)
            }
        }

        return super.findSnapView(layoutManager)
    }

    private fun findEndView(
        layoutManager: RecyclerView.LayoutManager,
        helper: OrientationHelper,
    ): View? {

        if (layoutManager is LinearLayoutManager) {

            val lastChild = layoutManager.findLastVisibleItemPosition()

            if (lastChild == RecyclerView.NO_POSITION) {
                return null
            }

            val child = layoutManager.findViewByPosition(lastChild)

            @Suppress("CascadeIf")
            return if (helper.getDecoratedStart(child) + helper.getDecoratedMeasurement(child) / 2 <= helper.totalSpace) {
                child
            } else if (layoutManager.findFirstCompletelyVisibleItemPosition() == 0) {
                null
            } else {
                layoutManager.findViewByPosition(lastChild - 1)
            }
        }

        return super.findSnapView(layoutManager)
    }

    private fun getVerticalHelper(layoutManager: RecyclerView.LayoutManager): OrientationHelper {
        var verticalHelper = this.verticalHelper
        if (verticalHelper == null) {
            verticalHelper =
                OrientationHelper.createVerticalHelper(layoutManager) as OrientationHelper
            this.verticalHelper = verticalHelper
        }
        return verticalHelper
    }

    private fun getHorizontalHelper(layoutManager: RecyclerView.LayoutManager): OrientationHelper {
        var horizontalHelper = this.horizontalHelper
        if (horizontalHelper == null) {
            horizontalHelper =
                OrientationHelper.createHorizontalHelper(layoutManager) as OrientationHelper
            this.horizontalHelper = horizontalHelper
        }
        return horizontalHelper
    }

    // var columnWidth : Int = 0

    override fun findTargetSnapPosition(
        layoutManager: RecyclerView.LayoutManager,
        velocityX: Int,
        velocityY: Int,
    ): Int {

        var targetPos = super.findTargetSnapPosition(layoutManager, velocityX, velocityY)
        if (targetPos != RecyclerView.NO_POSITION) {
            val currentView = findSnapView(layoutManager)
            if (currentView != null) {
                val currentPosition = layoutManager.getPosition(currentView)

                val clip = 1

                if (targetPos - currentPosition > clip) {
                    targetPos = currentPosition + clip
                } else if (targetPos - currentPosition < -clip) {
                    targetPos = currentPosition - clip
                }
            }
        }
        return targetPos
    }
}
